File: server/webapp/WEB-INF/rails.new/app/presenters/api_v1/config/tracking_tool/tracking_tool_representer.rb
##########################################################################
# Copyright 2015 ThoughtWorks, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
##########################################################################

module ApiV1
  module Config
    module TrackingTool
      class TrackingToolRepresenter < ApiV1::BaseRepresenter
        TRACKING_TOOL_TYPE_TO_REPRESENTER_MAP = {
          com.thoughtworks.go.config.MingleConfig => MingleTrackingToolRepresenter,
          com.thoughtworks.go.config.TrackingTool => ExternalTrackingToolRepresenter
        }

        TRACKING_TOOL_TYPE_TO_CLASS_MAP = {
          'generic' => com.thoughtworks.go.config.TrackingTool,
          'mingle'  => com.thoughtworks.go.config.MingleConfig
        }.freeze

        TRACKING_TOOL_CLASS_TO_TYPE_MAP = TRACKING_TOOL_TYPE_TO_CLASS_MAP.invert.freeze

        alias_method :tracking_tool, :represented
        property :type, exec_context: :decorator, skip_parse: true
        nested :attributes,
               decorator: lambda { |tracking_tool, *|
                 TRACKING_TOOL_TYPE_TO_REPRESENTER_MAP[tracking_tool.class]
               }

        property :errors, exec_context: :decorator, decorator: ApiV1::Config::ErrorRepresenter, skip_parse: true, skip_render: lambda { |object, options| object.empty? }


        def errors
          mapped_errors = {}
          tracking_tool.errors.each do |key, value|
            mapped_errors[matching_error_key(key)] = value
          end
          mapped_errors
        end

        class << self
          def get_class(type)
            TRACKING_TOOL_TYPE_TO_CLASS_MAP[type] || (raise ApiV1::UnprocessableEntity, "Invalid Tracking Tool type '#{type}'. It has to be one of '#{TRACKING_TOOL_TYPE_TO_CLASS_MAP.keys.join(', ')}.'")
          end
        end

        private

        def error_keys
          tool_class = TRACKING_TOOL_TYPE_TO_REPRESENTER_MAP[tracking_tool.class]
          tool_class.new(tracking_tool).error_keys
        end

        def matching_error_key key
          return error_keys[key] if error_keys[key]
          key
        end

        def type
          TRACKING_TOOL_CLASS_TO_TYPE_MAP[tracking_tool.class] || (raise ApiV1::UnprocessableEntity, "Invalid Tracking Tool type '#{tracking_tool.class}'. It has to be one of '#{TRACKING_TOOL_CLASS_TO_TYPE_MAP.keys.join(', ')}.'")
        end
      end
    end
  end
end

File: api/api-pipeline-config-v7/src/main/java/com/thoughtworks/go/apiv7/admin/pipelineconfig/representers/trackingtool/TrackingToolRepresenter.java
/*
 * Copyright 2018 ThoughtWorks, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.thoughtworks.go.apiv7.admin.pipelineconfig.representers.trackingtool;

import com.thoughtworks.go.api.base.OutputWriter;
import com.thoughtworks.go.api.representers.ErrorGetter;
import com.thoughtworks.go.api.representers.JsonReader;
import com.thoughtworks.go.config.MingleConfig;
import com.thoughtworks.go.config.PipelineConfig;
import com.thoughtworks.go.config.TrackingTool;
import com.thoughtworks.go.config.exceptions.UnprocessableEntityException;

import java.util.HashMap;

public class TrackingToolRepresenter {

  public static void toJSON(OutputWriter jsonWriter, PipelineConfig pipelineConfig) {
    HashMap<String, String> mapping = new HashMap<>();
    mapping.put("projectIdentifier", "project_identifier");
    mapping.put("baseUrl", "base_url");
    mapping.put("link", "url_pattern");
    if (pipelineConfig.getMingleConfig().isDefined()) {
      MingleConfig mingleConfig = pipelineConfig.getMingleConfig();
      if (!mingleConfig.errors().isEmpty()) {
        jsonWriter.addChild("errors", errorWriter -> {
          new ErrorGetter(mapping).toJSON(errorWriter, mingleConfig);
        });
      }
      jsonWriter.add("type", "mingle");
      jsonWriter.addChild("attributes", attributeWriter -> MingleTrackingToolRepresenter.toJSON(attributeWriter, mingleConfig));

    } else if (pipelineConfig.getTrackingTool() != null) {
      TrackingTool trackingTool = pipelineConfig.getTrackingTool();
      if (!trackingTool.errors().isEmpty()) {
        jsonWriter.addChild("errors", errorWriter -> {
          new ErrorGetter(mapping).toJSON(errorWriter, trackingTool);
        });
      }
      jsonWriter.add("type", "generic");
      jsonWriter.addChild("attributes", attributeWriter -> ExternalTrackingToolRepresenter.toJSON(attributeWriter, trackingTool));
    }

  }

  public static Object fromJSON(JsonReader jsonReader) {
    String type = jsonReader.getString("type");
    JsonReader attributes = jsonReader.readJsonObject("attributes");
    switch (type) {
      case "generic":
        return ExternalTrackingToolRepresenter.fromJSON(attributes);
      case "mingle":
        return MingleTrackingToolRepresenter.fromJSON(attributes);
      default:
        throw new UnprocessableEntityException(String.format("Invalid Tracking tool type '%s'. It has to be one of '%s'.", type, String.join(", ", "generic", "mingle")));
    }
  }
}
